// @prismicio/client/dist/helpers/asHTML.js
import { serializeEmbed, serializeHyperlink, serializeImage, serializePreFormatted, serializeSpan, serializeStandardTag } from "../lib/serializerHelpers.js";
import { serialize } from "../richtext/serialize.js";
import { wrapMapSerializer } from "../richtext/wrapMapSerializer.js";
import { composeSerializers } from "../richtext/composeSerializers.js";

//#region src/helpers/asHTML.ts
/**
* Creates a HTML rich text serializer with a given link resolver and provide
* sensible and safe defaults for every node type
*
* @internal
*/
const createHTMLRichTextSerializer = (linkResolver, serializer) => {
	const useSerializerOrDefault = (nodeSerializerOrShorthand, defaultWithShorthand) => {
		if (typeof nodeSerializerOrShorthand === "function") return ((payload) => {
			return (nodeSerializerOrShorthand === null || nodeSerializerOrShorthand === void 0 ? void 0 : nodeSerializerOrShorthand(payload)) || defaultWithShorthand(payload);
		});
		return defaultWithShorthand;
	};
	return wrapMapSerializerWithStringChildren({
		heading1: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.heading1, serializeStandardTag("h1", serializer === null || serializer === void 0 ? void 0 : serializer.heading1)),
		heading2: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.heading2, serializeStandardTag("h2", serializer === null || serializer === void 0 ? void 0 : serializer.heading2)),
		heading3: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.heading3, serializeStandardTag("h3", serializer === null || serializer === void 0 ? void 0 : serializer.heading3)),
		heading4: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.heading4, serializeStandardTag("h4", serializer === null || serializer === void 0 ? void 0 : serializer.heading4)),
		heading5: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.heading5, serializeStandardTag("h5", serializer === null || serializer === void 0 ? void 0 : serializer.heading5)),
		heading6: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.heading6, serializeStandardTag("h6", serializer === null || serializer === void 0 ? void 0 : serializer.heading6)),
		paragraph: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.paragraph, serializeStandardTag("p", serializer === null || serializer === void 0 ? void 0 : serializer.paragraph)),
		preformatted: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.preformatted, serializePreFormatted(serializer === null || serializer === void 0 ? void 0 : serializer.preformatted)),
		strong: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.strong, serializeStandardTag("strong", serializer === null || serializer === void 0 ? void 0 : serializer.strong)),
		em: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.em, serializeStandardTag("em", serializer === null || serializer === void 0 ? void 0 : serializer.em)),
		listItem: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.listItem, serializeStandardTag("li", serializer === null || serializer === void 0 ? void 0 : serializer.listItem)),
		oListItem: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.oListItem, serializeStandardTag("li", serializer === null || serializer === void 0 ? void 0 : serializer.oListItem)),
		list: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.list, serializeStandardTag("ul", serializer === null || serializer === void 0 ? void 0 : serializer.list)),
		oList: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.oList, serializeStandardTag("ol", serializer === null || serializer === void 0 ? void 0 : serializer.oList)),
		image: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.image, serializeImage(linkResolver, serializer === null || serializer === void 0 ? void 0 : serializer.image)),
		embed: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.embed, serializeEmbed(serializer === null || serializer === void 0 ? void 0 : serializer.embed)),
		hyperlink: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.hyperlink, serializeHyperlink(linkResolver, serializer === null || serializer === void 0 ? void 0 : serializer.hyperlink)),
		label: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.label, serializeStandardTag("span", serializer === null || serializer === void 0 ? void 0 : serializer.label)),
		span: useSerializerOrDefault(serializer === null || serializer === void 0 ? void 0 : serializer.span, serializeSpan())
	});
};
/**
* Wraps a map serializer into a regular function serializer. The given map
* serializer should accept children as a string, not as an array of strings
* like `@prismicio/client/richtext`'s `wrapMapSerializer`.
*
* @param mapSerializer - Map serializer to wrap
*
* @returns A regular function serializer
*/
const wrapMapSerializerWithStringChildren = (mapSerializer) => {
	const modifiedMapSerializer = {};
	for (const tag in mapSerializer) {
		const tagSerializer = mapSerializer[tag];
		if (tagSerializer) modifiedMapSerializer[tag] = (payload) => {
			return tagSerializer({
				...payload,
				children: payload.children.join("")
			});
		};
	}
	return wrapMapSerializer(modifiedMapSerializer);
};
const asHTML = (richTextField, ...configObjectOrTuple) => {
	if (richTextField) {
		const [configObjectOrLinkResolver, maybeSerializer] = configObjectOrTuple;
		let config;
		if (typeof configObjectOrLinkResolver === "function" || configObjectOrLinkResolver == null) config = {
			linkResolver: configObjectOrLinkResolver,
			serializer: maybeSerializer
		};
		else config = { ...configObjectOrLinkResolver };
		let serializer;
		if (config.serializer) if (typeof config.serializer === "function") serializer = composeSerializers((type, node, text, children, key) => config.serializer(type, node, text, children.join(""), key), createHTMLRichTextSerializer(config.linkResolver));
		else serializer = createHTMLRichTextSerializer(config.linkResolver, config.serializer);
		else serializer = createHTMLRichTextSerializer(config.linkResolver);
		return serialize(richTextField, serializer).join("");
	} else return null;
};

//#endregion
export { asHTML };
