// @prismicio/client/dist/index.cjs
const require_richText = require('./types/value/richText.cjs');
const require_mapSliceZone = require('./helpers/mapSliceZone.cjs');
const require_filter = require('./filter.cjs');
const require_cookie = require('./cookie.cjs');
const require_errors = require('./errors.cjs');
const require_link = require('./types/value/link.cjs');
const require_documentToLinkField = require('./helpers/documentToLinkField.cjs');
const require_asLink = require('./helpers/asLink.cjs');
const require_buildQueryURL = require('./buildQueryURL.cjs');
const require_isRepositoryName = require('./isRepositoryName.cjs');
const require_getRepositoryEndpoint = require('./getRepositoryEndpoint.cjs');
const require_getRepositoryName = require('./getRepositoryName.cjs');
const require_isRepositoryEndpoint = require('./isRepositoryEndpoint.cjs');
const require_Client = require('./Client.cjs');
const require_createClient = require('./createClient.cjs');
const require_Asset = require('./types/migration/Asset.cjs');
const require_Document = require('./types/migration/Document.cjs');
const require_isFilled = require('./helpers/isFilled.cjs');
const require_WriteClient = require('./WriteClient.cjs');
const require_createWriteClient = require('./createWriteClient.cjs');
const require_Migration = require('./Migration.cjs');
const require_createMigration = require('./createMigration.cjs');
const require_getGraphQLEndpoint = require('./getGraphQLEndpoint.cjs');
const require_getToolbarSrc = require('./getToolbarSrc.cjs');
const require_asDate = require('./helpers/asDate.cjs');
const require_asLinkAttrs = require('./helpers/asLinkAttrs.cjs');
const require_asText = require('./helpers/asText.cjs');
const require_asHTML = require('./helpers/asHTML.cjs');
const require_asImageSrc = require('./helpers/asImageSrc.cjs');
const require_asImageWidthSrcSet = require('./helpers/asImageWidthSrcSet.cjs');
const require_asImagePixelDensitySrcSet = require('./helpers/asImagePixelDensitySrcSet.cjs');
const require_embed = require('./types/value/embed.cjs');
const require_types = require('./types/model/types.cjs');
const require_link$1 = require('./types/model/link.cjs');
const require_sliceZone = require('./types/model/sliceZone.cjs');
const require_slice = require('./types/model/slice.cjs');
const require_types$1 = require('./types/webhook/types.cjs');

//#region src/index.ts
/**
* @deprecated Renamed to `filter`
*/
const predicate = require_filter.filter;
/**
* @deprecated Renamed to `mapSliceZone`
*/
const unstable_mapSliceZone = require_mapSliceZone.mapSliceZone;
/**
* @deprecated Use {@link RichTextNodeType} instead.
*/
const Element = require_richText.RichTextNodeType;

//#endregion
exports.Client = require_Client.Client;
exports.CustomTypeModelFieldType = require_types.CustomTypeModelFieldType;
exports.CustomTypeModelLinkSelectType = require_link$1.CustomTypeModelLinkSelectType;
exports.CustomTypeModelSliceDisplay = require_slice.CustomTypeModelSliceDisplay;
exports.CustomTypeModelSliceType = require_sliceZone.CustomTypeModelSliceType;
exports.Element = Element;
exports.ForbiddenError = require_errors.ForbiddenError;
exports.InvalidDataError = require_errors.InvalidDataError;
exports.LinkType = require_link.LinkType;
exports.Migration = require_Migration.Migration;
exports.NotFoundError = require_errors.NotFoundError;
exports.OEmbedType = require_embed.OEmbedType;
exports.ParsingError = require_errors.ParsingError;
exports.PreviewTokenExpiredError = require_errors.PreviewTokenExpiredError;
exports.PrismicError = require_errors.PrismicError;
exports.PrismicMigrationAsset = require_Asset.PrismicMigrationAsset;
exports.PrismicMigrationDocument = require_Document.PrismicMigrationDocument;
exports.RefExpiredError = require_errors.RefExpiredError;
exports.RefNotFoundError = require_errors.RefNotFoundError;
exports.RepositoryNotFoundError = require_errors.RepositoryNotFoundError;
exports.RichTextNodeType = require_richText.RichTextNodeType;
exports.WebhookType = require_types$1.WebhookType;
exports.WriteClient = require_WriteClient.WriteClient;
exports.asDate = require_asDate.asDate;
exports.asHTML = require_asHTML.asHTML;
exports.asImagePixelDensitySrcSet = require_asImagePixelDensitySrcSet.asImagePixelDensitySrcSet;
exports.asImageSrc = require_asImageSrc.asImageSrc;
exports.asImageWidthSrcSet = require_asImageWidthSrcSet.asImageWidthSrcSet;
exports.asLink = require_asLink.asLink;
exports.asLinkAttrs = require_asLinkAttrs.asLinkAttrs;
exports.asText = require_asText.asText;
exports.buildQueryURL = require_buildQueryURL.buildQueryURL;
Object.defineProperty(exports, 'cookie', {
  enumerable: true,
  get: function () {
    return require_cookie.cookie_exports;
  }
});
exports.createClient = require_createClient.createClient;
exports.createMigration = require_createMigration.createMigration;
exports.createWriteClient = require_createWriteClient.createWriteClient;
exports.documentToLinkField = require_documentToLinkField.documentToLinkField;
exports.filter = require_filter.filter;
exports.getGraphQLEndpoint = require_getGraphQLEndpoint.getGraphQLEndpoint;
exports.getRepositoryEndpoint = require_getRepositoryEndpoint.getRepositoryEndpoint;
exports.getRepositoryName = require_getRepositoryName.getRepositoryName;
exports.getToolbarSrc = require_getToolbarSrc.getToolbarSrc;
Object.defineProperty(exports, 'isFilled', {
  enumerable: true,
  get: function () {
    return require_isFilled.isFilled_exports;
  }
});
exports.isRepositoryEndpoint = require_isRepositoryEndpoint.isRepositoryEndpoint;
exports.isRepositoryName = require_isRepositoryName.isRepositoryName;
exports.mapSliceZone = require_mapSliceZone.mapSliceZone;
exports.predicate = predicate;
exports.unstable_mapSliceZone = unstable_mapSliceZone;
